Tag images in ECR with latest, in addition to the versioned tag

pvallone opened the pull request:
**What this PR does / why we need it**:
To facilitate automatically deploying new versions of our fork to `enterprise-dev`, we need to additionally tag our docker images with `latest`. When shipping a release to customers, we'll use the versioned tag that includes the commit hash. When deploying to `enterprise-dev`, we'll just use `latest`.

**Special notes for your reviewer**:
The `ECRPushImage` task only supports specifying one tag, so we need to push twice to add both tags. This seems to be the only way to push multiple tags to an image. This does not create multiple images.

Note that our other pipelines use the `Docker@2` task, which does support specifying multiple tags when pushing. I tried to use that task here, but I ran into permission problems. I assume that's why this GitHub pipeline was already using the `ECRPushImage` task.

Verified that both tags show up on ECR (I deleted the test image afterwards):
![image](https://user-images.githubusercontent.com/4745977/154873168-801bd100-d808-40f6-8599-3c5c262b6935.png)

Note also that the second push completes almost instantly:
![image](https://user-images.githubusercontent.com/4745977/154873195-374adf64-1d2f-4159-8256-e85b1491554c.png)

pvallone committed:
WIP

pvallone committed:
WIP without a condition for testing

pvallone committed:
I guess we'll use ECR push image

pvallone committed:
what if we only do one push

pvallone committed:
Revert "what if we only do one push"

This reverts commit 2276ed59c098bb964e34fcd68f742dfea1c8c81b.

pvallone committed:
Add back condition. Add comments

kjohn1922 commented:
I'd appreciate some implementation details in the description. The second ECRPushImage looks concerning without the context that ECRPushImage can only push one tag at a time (and that it doesn't create a second image, although that can be deduced from your screenshot). Looks like this is the expected way to use ECRPushImage to add multiple tags, but it took a bit of digging to find that information.

pvallone committed:
Add more context to comment justifying the second push

pvallone commented:
> I'd appreciate some implementation details in the description. The second ECRPushImage looks concerning without the context that ECRPushImage can only push one tag at a time (and that it doesn't create a second image, although that can be deduced from your screenshot). Looks like this is the expected way to use ECRPushImage to add multiple tags, but it took a bit of digging to find that information.

Sure, I've updated the description. I also added some more context to the comment describing the second push.

cameronwaterman commented on azure-pipelines.yml:
```diff
@@ -39,6 +41,22 @@ steps:
+# ECRPushImage does not support specifying multiple tags at push time. We therefore push a second time to
+# additionally tag the image with the 'latest' tag. Note that this does not create a second image, and
+# this push completes very quickly since it just adds an additional tag to the image.
+- task: ECRPushImage@1
```
Maybe this is the reason we aren't using this task in our other pipelines. Do you think it would be more clear to perform the image push the same way we do in our other pipelines?

https://ni.visualstudio.com/DefaultCollection/DevCentral/_git/Skyline?path=/Web/Workspaces/Grafana/azure-pipeline.yml&version=GBmaster&line=210&lineEnd=211&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

pvallone replied:
Yeah, I think we shouldn't be different unless we have a reason to be. I don't know why this pipeline is currently using `ECRPushImage`. I didn't realize this pipeline was unique in that respect, or I would have raised an issue when the pipeline was originally created. In this case, I couldn't get the task we use in our other pipelines to work (see [this error here when I tried to use the Docker2 task](https://ni.visualstudio.com/DevCentral/_build/results?buildId=2197057&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=0febf08e-6fac-5ca3-6c4b-78e0cd0a0a95&l=44)). Perhaps that's why we're using `ECRPushImage`?

cameronwaterman replied:
I talked with Carson about this previously. He used this task because it seemed like a more first-class way of pushing to ECR, but acknowledged the limitation regarding multiple tags. I'm assuming the other task is failing because we need to login first. If you look at the pipeline I linked there are some steps prior to the push you probably need to pull in as well.

pvallone replied:
Gotcha, I'll try adding the login tasks.

pvallone replied:
Unfortunately it [still fails](https://ni.visualstudio.com/DevCentral/_build/results?buildId=2201186&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=347de3ba-75e7-5260-17f8-800303757031&l=41) even after adding the docker hub login tasks. I'm guessing it has something to do with how we're not running these tasks on the skyline build agents, but are instead using the DevCentral-Docker agents. I suspect it isn't a tricky thing to fix, but given our schedule constraints, I'd like to propose just sticking to the `ECRPushImage` task that the pipeline is already using.

I think it makes sense create some tech debt for making this pipeline look like our other pipelines and replacing the `ECRPushImage` tasks with what we do everywhere else.

cameronwaterman replied:
Works for me. I agree, it likely is because we're in a different agent pool. I'll go ahead and count this as resolved.

pvallone committed:
Attempting to use the native Docker tasks

pvallone committed:
Revert "Attempting to use the native Docker tasks"

This reverts commit b5a641beb2a4c07149171c722ef9908b0148c495.
